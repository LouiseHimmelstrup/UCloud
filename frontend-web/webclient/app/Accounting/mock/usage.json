{
    "data": {
        "usage": 1333333337,
        "dataType": "bytes",
        "quota": 13333333337,
        "title": "Storage Used"
    }
}