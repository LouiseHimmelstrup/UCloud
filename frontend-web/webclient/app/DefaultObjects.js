import {Cloud} from "../authentication/SDUCloudObject"
import {tusConfig} from "./Configurations";
import Uppy from "uppy";

export const DefaultStatus = {
    title: "No Issues",
    level: "NO ISSUES",
    body: "The system is running as intended.",
};

export const RightsMap = {
    NONE: 0,
    READ: 1,
    READ_WRITE: 2,
    OWN: 3,
};

export const AnalysesStatusMap = {
    "PENDING": 0,
    "IN PROGRESS": 1,
    "COMPLETED": 2,
};

export const RightsNameMap = {
    NONE: "None",
    READ: "Read",
    READ_WRITE: "Read/Write",
    OWN: "Own",
};

export const SensitivityLevel = {
    "OPEN_ACCESS": "Open Access",
    "CONFIDENTIAL": "Confidential",
    "SENSITIVE": "Sensitive",
};

export const SensitivityLevelMap = {
    "OPEN_ACCESS": 0,
    "CONFIDENTIAL": 1,
    "SENSITIVE": 2,
};

const initializeUppy = (restrictions) => 
    Uppy.Core({
        autoProceed: false,
        debug: false,
        restrictions: restrictions,
        meta: {
            sensitive: false,
        },
        onBeforeUpload: () => {
            return Cloud.receiveAccessTokenOrRefreshIt().then((data) => {
                tusConfig.headers["Authorization"] = `Bearer ${data}`;
            });
        }
    }).use(Uppy.Tus, tusConfig);

export const initObject = {
    files: {
        files: [],
        filesShown: 10,
        filesPerPage: 10,
        currentFilesPage: 0,
        loading: false,
        path: "",
        filesInfoPath: "",
        filesInfoFile: null,
        projects: []
    },
    uppy: {
        uppyFiles: initializeUppy({maxNumberOfFiles: false}),
        uppyFilesOpen: false,
        uppyRunApp: initializeUppy({ maxNumberOfFiles: 1}),
        uppyRunAppOpen: false
    },
    status: {
        status: DefaultStatus,
        title: ""
    }
};