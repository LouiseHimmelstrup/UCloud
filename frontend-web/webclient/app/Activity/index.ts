import { ActivityReduxObject Â } from "DefaultObjects";

export type Activity = CountedActivity | TrackedActivity;

export type CountedOperations = "FAVORITE" | "DOWNLOAD";
export type TrackedOperations = "CREATE" | "UPDATE" | "DELETE" | "MOVED";

interface CountedActivityEntry {
    id: string
    path: string | null
    count: number
}

export interface CountedActivity {
    type: "counted"
    operation: CountedOperations
    entries: CountedActivityEntry[]
    timestamp: number
}

interface TrackedActivityFile {
    id: string
    path: string
}

export interface TrackedActivity {
    type: "tracked"
    operation: TrackedOperations
    files: TrackedActivityFile[]
    timestamp: number
}

export interface ActivityDispatchProps {
    fetchActivity: (pageNumber: number, pageSize: number) => void
    setError: (error?: string) => void
    setPageTitle: () => void
}

export interface ActivityProps extends ActivityReduxObject, ActivityDispatchProps { }