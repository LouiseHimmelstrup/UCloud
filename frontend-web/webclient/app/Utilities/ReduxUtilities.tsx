import dashboard from "Dashboard/Redux/DashboardReducer";
import {initObject} from "DefaultObjects";
import header, {CONTEXT_SWITCH, USER_LOGIN, USER_LOGOUT} from "Navigation/Redux/HeaderReducer";
import sidebar from "Navigation/Redux/SidebarReducer";
import status from "Navigation/Redux/StatusReducer";
import notifications from "Notifications/Redux/NotificationsReducer";
import * as ProjectRedux from "Project/Redux";
import {Action, AnyAction, combineReducers, createStore, Store} from "redux";
import {composeWithDevTools} from "redux-devtools-extension";
import {createResponsiveStateReducer, responsiveStoreEnhancer} from "redux-responsive";
import simpleSearch from "Search/Redux/SearchReducer";
import {responsiveBP} from "ui-components/theme";
import avatar from "UserSettings/Redux/AvataaarReducer";
import hookStore from "Utilities/ReduxHooks";

export function configureStore(
    initialObject: ReduxObject,
    reducers,
    enhancers?
): Store<ReduxObject, AnyAction> {
    const combinedReducers = combineReducers<ReduxObject, AnyAction>(reducers);
    const rootReducer = (state: ReduxObject, action: Action): ReduxObject => {
        if ([USER_LOGIN, USER_LOGOUT, CONTEXT_SWITCH].some(it => it === action.type)) {
            state = initObject();
        }
        return combinedReducers(state, action);
    };
    return createStore<ReduxObject, AnyAction, {}, {}>(rootReducer, initialObject, composeWithDevTools(enhancers));
}

export const responsive = createResponsiveStateReducer(
    responsiveBP,
    {infinity: "xxl"}
);

export const store = configureStore(initObject(), {
    dashboard,
    header,
    status,
    sidebar,
    notifications,
    simpleSearch,
    hookStore,
    avatar,
    loading,
    project: ProjectRedux.reducer,
    responsive: createResponsiveStateReducer(
        responsiveBP,
        {infinity: "xxl"}),
}, responsiveStoreEnhancer);

function loading(state = false, action: {type: string}): boolean {
    switch (action.type) {
        case "LOADING_START":
            return true;
        case "LOADING_END":
            return false;
        default:
            return state;
    }
}
