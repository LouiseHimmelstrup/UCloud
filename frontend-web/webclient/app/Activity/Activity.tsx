import * as React from "react";
import { connect } from "react-redux";
import { GroupedActivity, ActivityProps, ActivityDispatchProps } from "Activity";
import * as Module from "Activity";
import * as Pagination from "Pagination";
import * as moment from "moment";
import { getFilenameFromPath, replaceHomeFolder } from "Utilities/FileUtilities";
import { ActivityReduxObject, ReduxObject } from "DefaultObjects";
import { fetchActivity, setErrorMessage, setLoading } from "./Redux/ActivityActions";
import { updatePageTitle, setActivePage } from "Navigation/Redux/StatusActions";
import { Dispatch } from "redux";
import { fileInfoPage } from "Utilities/FileUtilities";
import * as Heading from "ui-components/Heading"
import Icon, { IconName } from "ui-components/Icon";
import { Flex, Text, Link } from "ui-components";
import Table, { TableRow, TableCell, TableBody, TableHeader, TableHeaderCell } from "ui-components/Table";
import { Dropdown, DropdownContent } from "ui-components/Dropdown";
import { Cloud } from "Authentication/SDUCloudObject";
import { MainContainer } from "MainContainer/MainContainer";
import styled from "styled-components";
import { SidebarPages } from "ui-components/Sidebar";
import { setRefreshFunction } from "Navigation/Redux/HeaderActions";
import { Spacer } from "ui-components/Spacer";

class Activity extends React.Component<ActivityProps> {

    public componentDidMount() {
        this.props.setPageTitle();
        this.props.setActivePage();
        this.props.fetchActivity(0, 100);
    }

    public componentWillUnmount() {
        this.props.clearRefresh();
    }

    public render() {
        const { fetchActivity, page, error, setError, loading, groupedEntries } = this.props;

        const main = (
            <React.StrictMode>
                <Pagination.List
                    loading={loading}
                    errorMessage={error}
                    onErrorDismiss={setError}
                    pageRenderer={() => <ActivityFeedGrouped activity={groupedEntries ? groupedEntries : []} />}
                    page={page}
                    onPageChanged={pageNumber => fetchActivity(pageNumber, page.itemsPerPage)}
                />
            </React.StrictMode>
        );

        const header = (<Spacer left={<Heading.h2>File Activity</Heading.h2>} right={
            <Pagination.EntriesPerPageSelector
                onChange={itemsPerPage => fetchActivity(page.pageNumber, itemsPerPage)}
                content="Activity per page"
                entriesPerPage={page.itemsPerPage}
            />
        } />);

        return (
            <MainContainer
                main={main}
                header={header}
            />
        );
    }
}


const ActivityFeedGrouped = ({ activity }: { activity: GroupedActivity[] }) => activity.length ? (
    <Table>
        <TableHeader>
            <TFRow>
                <TableHeaderCell width="7em" />
                <TableHeaderCell width="10.5em" />
                <TableHeaderCell width="99%" />
            </TFRow>
        </TableHeader>
        <TableBody>
            {activity.map((a, i) => <TrackedFeedActivity key={i} activity={a} />)}
        </TableBody>
    </Table>
) : null;

export const ActivityFeed = ({ activity }: { activity: Module.Activity[] }) =>
    <ActivityFeedGrouped activity={groupActivity(activity)!} />;

const ActivityOperation: React.FunctionComponent<{ event: Module.Activity }> = props => {
    switch (props.event.type) {
        case Module.ActivityType.MOVED: {
            return <>
                was moved to
                {" "}
                <b><Link to={fileInfoPage((props.event as Module.MovedActivity).newName)}>
                    {replaceHomeFolder((props.event as Module.MovedActivity).newName, Cloud.homeFolder)}
                </Link></b>
            </>;
        }

        case Module.ActivityType.FAVORITE: {
            const isFavorite = (props.event as Module.FavoriteActivity).favorite;
            if (isFavorite) {
                return <>was <b>added to favorites</b></>;
            } else {
                return <>was <b>removed from favorites</b></>;
            }
        }

        default: {
            return <>was <b>{operationToPastTense(props.event.type)}</b></>;
        }
    }
};

const ActivityEvent: React.FunctionComponent<{ event: Module.Activity }> = props => (
    <Text fontSize={1}>
        <b>
            <Link to={fileInfoPage(props.event.originalFilePath)}>
                {getFilenameFromPath(props.event.originalFilePath)}
            </Link>
        </b> <ActivityOperation event={props.event} />
    </Text>

);

const TrackedFeedActivity = ({ activity }: { activity: GroupedActivity }) => (
    <TFRow>
        <TableCell>
            <Dropdown>
                <Text fontSize={1} color="text">{moment(activity.timestamp).fromNow()}</Text>
                <DropdownContent>
                    {moment(activity.timestamp).format("llll")}
                </DropdownContent>
            </Dropdown>
        </TableCell>
        <TableCell>
            <Flex>
                <Icon mr="0.5em" name={eventIcon(activity.type).icon} />
                <Text fontSize={2}>{`Files ${operationToPastTense(activity.type)}`}</Text>
            </Flex>
        </TableCell>
        <TableCell>
            {activity.entries.map((item, idx) =>
                <ActivityEvent key={idx} event={item} />
            )}
        </TableCell>
    </TFRow>
);

const operationToPastTense = (operation: Module.ActivityType): string => {
    switch (operation) {
        case Module.ActivityType.DELETED: return "deleted";
        case Module.ActivityType.DOWNLOAD: return "downloaded";
        case Module.ActivityType.FAVORITE: return "favorited";
        case Module.ActivityType.INSPECTED: return "inspected";
        case Module.ActivityType.MOVED: return "moved";
        case Module.ActivityType.UPDATED: return "updated";
    }
}

interface EventIconAndColor {
    icon: IconName,
    color: "blue" | "green" | "red",
    rotation?: 45
}

const eventIcon = (operation: Module.ActivityType): EventIconAndColor => {
    switch (operation) {
        case Module.ActivityType.FAVORITE:
            return { icon: "starFilled", color: "blue" };
        case Module.ActivityType.DOWNLOAD:
            return { icon: "download", color: "blue" };
        case Module.ActivityType.UPDATED:
            return { icon: "refresh", color: "green" };
        case Module.ActivityType.DELETED:
            return { icon: "close", color: "red" };
        case Module.ActivityType.MOVED:
            return { icon: "move", color: "green" };
        default:
            return { icon: "ellipsis", color: "blue" }
    }
}

function groupActivity(items: Module.Activity[] = []): GroupedActivity[] {
    const result: GroupedActivity[] = [];
    let currentGroup: GroupedActivity | null = null;

    const pushGroup = () => {
        if (currentGroup != null) {
            result.push(currentGroup);
            currentGroup = null;
        }
    };

    const initializeGroup = (item: Module.Activity) => {
        currentGroup = {
            type: item.type,
            timestamp: item.timestamp,
            entries: [item]
        };
    };

    items.forEach(item => {
        if (currentGroup === null) {
            initializeGroup(item);
        } else {
            if (currentGroup.type !== item.type || Math.abs(item.timestamp - currentGroup.timestamp) > (1000 * 60 * 15)) {
                pushGroup();
                initializeGroup(item);
            } else {
                currentGroup.entries.push(item);
            }
        }
    });

    pushGroup();
    return result;
}

const TFRow = styled(TableRow)`
    vertical-align: top;
`;

const mapStateToProps = ({ activity }: ReduxObject): ActivityReduxObject & Module.ActivityOwnProps => ({
    ...activity,
    groupedEntries: groupActivity(activity.page.items)
});

const mapDispatchToProps = (dispatch: Dispatch): ActivityDispatchProps => ({
    fetchActivity: (pageNumber, pageSize) => {
        const fetch = async () => {
            dispatch(setLoading(true));
            dispatch(await fetchActivity(pageNumber, pageSize));
        };
        fetch();
        dispatch(setRefreshFunction(fetch))
    },
    setError: error => dispatch(setErrorMessage(error)),
    setPageTitle: () => dispatch(updatePageTitle("Activity")),
    setActivePage: () => dispatch(setActivePage(SidebarPages.Activity)),
    clearRefresh: () => dispatch(setRefreshFunction())
});

export default connect(mapStateToProps, mapDispatchToProps)(Activity);