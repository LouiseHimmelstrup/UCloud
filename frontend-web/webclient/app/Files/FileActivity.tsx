import * as React from "react";
import { connect } from "react-redux";
import PromiseKeeper from "PromiseKeeper";
import { Cloud } from "Authentication/SDUCloudObject";
import { emptyPage } from "DefaultObjects";
import { FileActivityProps, FileActivityState, Activity, TrackedActivity, CountedActivity, TrackedOperations, CountedOperations } from "Files";
import { Feed, Icon, Segment, Header } from "semantic-ui-react";
import { Page } from "Types";
import * as Pagination from "Pagination";
import * as moment from "moment";
import { getFilenameFromPath } from "Utilities/FileUtilities";
import { Link } from "react-router-dom";

class FileActivity extends React.Component<FileActivityProps, FileActivityState> {
    constructor(props) {
        super(props)
        this.state = {
            promises: new PromiseKeeper(),
            activity: emptyPage
        }
    }

    componentDidMount() {
        this.fetchActivity(0, 25);
    }

    fetchActivity = (pageNumber: number = 0, pageSize: number = 25) => {
        this.state.promises.makeCancelable(
            Cloud.get(`/activity/stream?page=${pageNumber}&itemsPerPage=${pageSize}`)
        ).promise.then(({ response }) => this.setState(() => ({ activity: response })));
    }

    componentWillUnmount() {
        this.state.promises.cancelPromises();
    }

    render() {
        return (
            <React.StrictMode>
                <Header as="h1" content="File Activity" />
                <Pagination.List
                    pageRenderer={(page: Page<Activity>) => <Segment><ActivityFeed activity={page.items} /></Segment>}
                    page={this.state.activity}
                    onItemsPerPageChanged={() => undefined}
                    onPageChanged={() => undefined}
                />
            </React.StrictMode>
        );
    }
}

export const ActivityFeed = ({ activity }: { activity: Activity[] }) => activity.length ? (
    <Feed>
        {activity.map((a, i) => {
            if (a.type === "tracked") {
                return <TrackedFeedActivity key={i} activity={a} />
            } else if (a.type === "counted") {
                return <CountedFeedActivity key={i} activity={a} />
            } else { return null }
        })}
    </Feed>
) : null;

const CountedFeedActivity = ({ activity }: { activity: CountedActivity }) => (
    <Feed.Event
        icon={eventIcon2(activity.operation)}
        date={moment(new Date(activity.timestamp)).fromNow()}
        summary={`Files ${operationToPastTense(activity.operation)}`}
        extraText={activity.entries.map((entry, i) => !!entry.path ?
            (<div key={i}>
                <b><Link to={`/fileInfo/${entry.path}`}>{getFilenameFromPath(entry.path)}</Link></b> was <b>{operationToPastTense(activity.operation)}</b> {entry.count === 1 ? "once" : <><b>{entry.count}</b> times</>}</div>) : null
        )}
    />
);

const TrackedFeedActivity = ({ activity }: { activity: TrackedActivity }) => (
    <Feed.Event>
        <Feed.Label>
            <EventIcon operation={activity.operation} />
        </Feed.Label>
        <Feed.Content>
            <Feed.Summary>
                Files <b>{activity.files.join(", ")}</b> {activity.files.length > 1 ? "were" : "was"} {operationToPastTense(activity.operation)}
            </Feed.Summary>
            <Feed.Meta>
                {moment(new Date(activity.timestamp)).fromNow()}
            </Feed.Meta>
        </Feed.Content>
    </Feed.Event>
);

const operationToPastTense = (operation: TrackedOperations | CountedOperations) => {
    if (operation === "MOVED") return "moved";
    if ((operation as string).endsWith("E")) return `${(operation as string).toLowerCase()}d`;
    return `${operation}ed`;
}

const EventIcon = ({ operation }: { operation: TrackedOperations | CountedOperations }) => {
    switch (operation) {
        case "FAVORITE": {
            return <Icon name="favorite" />
        }
        case "DOWNLOAD": {
            return <Icon name="download" />
        }
        case "CREATE": {
            return <Icon name="plus" />
        }
        case "UPDATE": {
            return <Icon name="refresh" />
        }
        case "DELETE": {
            return <Icon name="delete" />
        }
        case "MOVED": {
            return <Icon name="move" />
        }
    }
}

const eventIcon2 = (operation: TrackedOperations | CountedOperations) => {
    switch (operation) {
        case "FAVORITE": {
            return "favorite";
        }
        case "DOWNLOAD": {
            return "download";
        }
        case "CREATE": {
            return "plus";
        }
        case "UPDATE": {
            return "refresh";
        }
        case "DELETE": {
            return "delete";
        }
        case "MOVED": {
            return "move";
        }
    }
}

export default connect()(FileActivity);