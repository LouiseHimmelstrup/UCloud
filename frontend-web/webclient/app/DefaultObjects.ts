import {SidebarOption, Page} from "Types";
import {Status} from "Navigation";
import {Analysis, DetailedApplicationSearchReduxState, RunsSortBy} from "Applications";
import {File, DetailedFileSearchReduxState} from "Files";
import {SortOrder, SortBy} from "Files";
import {DashboardStateProps} from "Dashboard";
import {Publication} from "Zenodo";
import {Notification} from "Notifications";
import {Upload} from "Uploader";
import {Activity, ActivityGroup, ActivityFilter} from "Activity";
import {Reducer} from "redux";
import {SimpleSearchStateProps} from "Search";
import * as ApplicationRedux from "Applications/Redux";
import * as AccountingRedux from "Accounting/Redux";
import * as SnackbarRedux from "Snackbar/Redux";
import * as FavoritesRedux from "Favorites/Redux";
import {defaultAvatar} from "UserSettings/Avataaar";
import {SidebarPages} from "ui-components/Sidebar";
import {ScrollResult} from "Scroll/Types";
import * as ProjectRedux from "Project/Redux";

export const DefaultStatus: Status = {
    title: "No Issues",
    level: "NO ISSUES",
    body: "The system is running as intended."
};

export enum KeyCode {
    ENTER = 13,
    ESC = 27
}

export const emptyPage: Page<any> = {items: [], itemsPerPage: 25, itemsInTotal: 0, pageNumber: 0, pagesInTotal: 0};

export enum AnalysesStatusMap {
    "PENDING",
    "IN PROGRESS",
    "COMPLETED"
}

export enum RightsNameMap {
    "NONE" = "None",
    "READ" = "Read",
    "READ_WRITE" = "Read/Write"
}

export enum SensitivityLevel {
    "INHERIT" = "Inherit",
    "PRIVATE" = "Private",
    "CONFIDENTIAL" = "Confidential",
    "SENSITIVE" = "Sensitive"
}

export type Sensitivity = keyof typeof SensitivityLevel;

export enum SensitivityLevelMap {
    INHERIT = "INHERIT",
    PRIVATE = "PRIVATE",
    CONFIDENTIAL = "CONFIDENTIAL",
    SENSITIVE = "SENSITIVE"
}

function getFilesSortingColumnOrDefault(columnIndex: 0 | 1): SortBy {
    const sortingColumn = window.localStorage.getItem(`filesSorting${columnIndex}`);
    if (sortingColumn && Object.values(SortBy).includes(sortingColumn)) return sortingColumn as SortBy;
    switch (columnIndex) {
        case 0:
            window.localStorage.setItem("filesSorting0", SortBy.MODIFIED_AT);
            return SortBy.MODIFIED_AT;
        case 1:
            window.localStorage.setItem("filesSorting1", SortBy.SIZE);
            return SortBy.SIZE;
    }
}

function getItemOrDefault<T, T2>(itemName: string, defaultValue: T, en: T2): T {
    const item = window.localStorage.getItem(itemName);
    if (item && Object.values(en).includes(item)) {
        return item as unknown as T;
    }
    return defaultValue;
}

export interface ComponentWithLoadingState {
    loading: boolean
    error?: string
}

export interface ComponentWithPage<T> extends ComponentWithLoadingState {
    page: Page<T>
}

export interface ComponentWithScroll<Item, OffsetType> extends ComponentWithLoadingState {
    scroll?: ScrollResult<Item, OffsetType>
}

export interface ResponsiveReduxObject {
    mediaType: string
    orientation: string
    lessThan: Record<string, boolean>
    greaterThan: Record<string, boolean>
    is: Record<string, boolean>
}

export const initResponsive = (): ResponsiveReduxObject => ({
    mediaType: "",
    orientation: "",
    lessThan: {},
    greaterThan: {},
    is: {},
});

export interface FilePreviewReduxState {
    file?: File
    error?: string
}

export interface FilesReduxObject extends ComponentWithPage<File> {
    sortOrder: SortOrder
    sortBy: SortBy
    path: string
    filesInfoPath: string
    fileSelectorError?: string
    sortingColumns: [SortBy, SortBy]
    fileSelectorLoading: boolean
    fileSelectorShown: boolean
    fileSelectorPage: Page<File>
    fileSelectorPath: string
    fileSelectorCallback: Function
    fileSelectorIsFavorites: boolean
    disallowedPaths: string[]
    invalidPath: boolean
}

export interface FileInfoReduxObject {
    file?: File
    error?: string
    activity: Page<Activity>
    loading: boolean
}

export interface AnalysisReduxObject extends ComponentWithPage<Analysis> {
    sortBy: RunsSortBy
    sortOrder: SortOrder
}

export interface NotificationsReduxObject {
    redirectTo: string
    items: Notification[]
    loading: boolean
    error?: string
}

export interface ZenodoReduxObject extends ComponentWithPage<Publication> {
    connected: boolean
}

export interface StatusReduxObject {
    status: Status
    title: string
    page: SidebarPages
    loading: boolean
}

export interface SidebarReduxObject {
    pp: boolean
    options: SidebarOption[]
    kcCount: number
}

export interface HeaderSearchReduxObject {
    prioritizedSearch: HeaderSearchType
    refresh?: () => void
}

export interface RunApplicationReduxObject {

}

export type ActivityReduxObject = ComponentWithScroll<ActivityGroup, number> & ActivityFilter;

export type HeaderSearchType = "files" | "applications" | "projects";

export interface UploaderReduxObject {
    uploads: Upload[]
    visible: boolean
    path: string
    allowMultiple: boolean
    onFilesUploaded: (p: string) => void
    error?: string
    loading: boolean
}

interface LegacyReducers {
    dashboard?: Reducer<DashboardStateProps>
    files?: Reducer<FilesReduxObject>
    uploader?: Reducer<UploaderReduxObject>
    status?: Reducer<StatusReduxObject>
    notifications?: Reducer<NotificationsReduxObject>
    analyses?: Reducer<AnalysisReduxObject>
    zenodo?: Reducer<ZenodoReduxObject>
    header?: Reducer<HeaderSearchReduxObject>
    sidebar?: Reducer<SidebarReduxObject>
    activity?: Reducer<ActivityReduxObject>
    detailedResult?: Reducer<DetailedResultReduxObject>
}

export type Reducers = LegacyReducers & ApplicationRedux.Reducers & AccountingRedux.Reducers;

export type DetailedResultReduxObject = ComponentWithPage<File>

export const initDetailedResult = (): DetailedResultReduxObject => ({
    page: emptyPage,
    loading: false,
    error: undefined
});

/* FIXME */
interface LegacyReduxObject {
    dashboard: DashboardStateProps
    files: FilesReduxObject,
    uploader: UploaderReduxObject
    status: StatusReduxObject,
    notifications: NotificationsReduxObject
    analyses: AnalysisReduxObject
    zenodo: ZenodoReduxObject
    header: HeaderSearchReduxObject
    sidebar: SidebarReduxObject
    activity: ActivityReduxObject
    detailedResult: DetailedResultReduxObject
    simpleSearch: SimpleSearchStateProps
    detailedFileSearch: DetailedFileSearchReduxState
    detailedApplicationSearch: DetailedApplicationSearchReduxState
    fileInfo: FileInfoReduxObject
    avatar: AvatarReduxObject
    filePreview: FilePreviewReduxState
    responsive?: ResponsiveReduxObject
    project: ProjectRedux.State
    loading?: boolean
}

export type ReduxObject =
    LegacyReduxObject &
    ApplicationRedux.Objects &
    AccountingRedux.Objects &
    FavoritesRedux.Objects &
    SnackbarRedux.Wrapper;

export const initActivity = (): ActivityReduxObject => ({
    loading: false
});

export const initNotifications = (): NotificationsReduxObject => ({
    items: [],
    loading: false,
    redirectTo: "",
    error: undefined
});

export const initHeader = (): HeaderSearchReduxObject => ({
    prioritizedSearch: "files"
});

export const initStatus = (): StatusReduxObject => ({
    status: DefaultStatus,
    title: "",
    page: SidebarPages.None,
    loading: false
});

export const initDashboard = (): DashboardStateProps => ({
    favoriteFiles: [],
    recentFiles: [],
    recentAnalyses: [],
    notifications: [],
    favoriteLoading: false,
    recentLoading: false,
    analysesLoading: false
});

export function initObject(homeFolder: string): ReduxObject {
    return {
        dashboard: initDashboard(),
        files: initFiles(homeFolder),
        status: initStatus(),
        header: initHeader(),
        notifications: initNotifications(),
        analyses: initAnalyses(),
        zenodo: initZenodo(),
        sidebar: initSidebar(),
        uploader: initUploads(),
        activity: initActivity(),
        detailedResult: initDetailedResult(),
        simpleSearch: initSimpleSearch(),
        detailedApplicationSearch: initApplicationsAdvancedSearch(),
        detailedFileSearch: initFilesDetailedSearch(),
        fileInfo: initFileInfo(),
        avatar: initAvatar(),
        filePreview: initFilePreview(),
        project: ProjectRedux.initialState,
        ...ApplicationRedux.init(),
        ...AccountingRedux.init(),
        ...FavoritesRedux.init(),
        ...SnackbarRedux.init(),
        responsive: undefined,
    }
}


export const initFilePreview = () => ({
    file: undefined,
    error: undefined
});

export type AvatarReduxObject = typeof defaultAvatar & { error?: string };
export const initAvatar = (): AvatarReduxObject => ({...defaultAvatar, error: undefined});

export const initSimpleSearch = (): SimpleSearchStateProps => ({
    files: emptyPage,
    filesLoading: false,
    applications: emptyPage,
    applicationsLoading: false,
    errors: [],
    search: "",
    applicationSearch: initApplicationsAdvancedSearch(),
    fileSearch: initFilesDetailedSearch()
});

export const initAnalyses = (): AnalysisReduxObject => ({
    page: emptyPage,
    loading: false,
    error: undefined,
    sortBy: RunsSortBy.createdAt,
    sortOrder: SortOrder.ASCENDING
});

export const initZenodo = (): ZenodoReduxObject => ({
    connected: false,
    loading: false,
    page: emptyPage
});

export const initSidebar = (): SidebarReduxObject => ({
    pp: false,
    kcCount: 0,
    options: []
});

export const initUploads = (): UploaderReduxObject => ({
    path: "",
    uploads: [],
    visible: false,
    allowMultiple: false,
    error: undefined,
    onFilesUploaded: () => null,
    loading: false
});

export const initFileInfo = (): FileInfoReduxObject => ({
    activity: emptyPage,
    loading: false
});

export const initFiles = (homeFolder: string): FilesReduxObject => ({
    page: emptyPage,
    sortOrder: getItemOrDefault("sortOrder", SortOrder.ASCENDING, SortOrder),
    sortBy: getItemOrDefault("sortBy", SortBy.PATH, SortBy),
    loading: false,
    error: undefined,
    path: "",
    filesInfoPath: "",
    sortingColumns: [getFilesSortingColumnOrDefault(0), getFilesSortingColumnOrDefault(1)],
    fileSelectorLoading: false,
    fileSelectorShown: false,
    fileSelectorPage: emptyPage,
    fileSelectorPath: homeFolder,
    fileSelectorIsFavorites: false,
    fileSelectorCallback: () => undefined,
    fileSelectorError: undefined,
    disallowedPaths: [],
    invalidPath: false
});

export const initFilesDetailedSearch = (): DetailedFileSearchReduxState => ({
    hidden: true,
    allowFolders: true,
    allowFiles: true,
    fileName: "",
    extensions: new Set(),
    tags: new Set(),
    sensitivities: new Set(),
    createdBefore: undefined,
    createdAfter: undefined,
    modifiedBefore: undefined,
    modifiedAfter: undefined,
    error: undefined,
    loading: false
});

export const initApplicationsAdvancedSearch = (): DetailedApplicationSearchReduxState => ({
    error: undefined,
    loading: false,
    hidden: true,
    appName: "",
    appVersion: "",
    tags: ""
});
