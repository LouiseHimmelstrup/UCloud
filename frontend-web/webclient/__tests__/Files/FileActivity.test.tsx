import * as React from "react";
import FileActivity, { ActivityFeed } from "Files/FileActivity";
import { create } from "react-test-renderer";
import { initAnalyses } from "DefaultObjects";
import analyses from "Applications/Redux/AnalysesReducer";
import { configureStore } from "Utilities/ReduxUtilities";
import { Provider } from "react-redux";
import { activityPage } from "../mock/Activity";
import { MemoryRouter } from "react-router";

describe("FileActivity", () => {
    test("Mount", () => {
        const store = configureStore({ analyses: initAnalyses() }, { analyses })
        expect(create(<Provider store={store}><FileActivity /></Provider>).toJSON()).toMatchSnapshot();
    });
});

describe("Activity Feed", () => {
    test("Mount, no activity", () => {
        expect(create(<ActivityFeed activity={[]} />))
    });

    test("Mount, with activity", () => {
        expect(create(<MemoryRouter><ActivityFeed activity={activityPage.items} /></MemoryRouter>))
    });
})