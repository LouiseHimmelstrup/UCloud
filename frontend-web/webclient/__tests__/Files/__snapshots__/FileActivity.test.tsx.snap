// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileActivity Mount 1`] = `
Array [
  <h1
    className="ui header"
  >
    File Activity
  </h1>,
  <span
    className="items-per-page-padding pagination-float-right"
  >
    <span>
      Items per page 
    </span>
    <div
      aria-expanded={false}
      className="ui compact inline dropdown"
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      role="listbox"
      tabIndex={0}
    >
      <div
        aria-live="polite"
        className="text"
        role="alert"
      >
        25
      </div>
      <i
        aria-hidden="true"
        className="dropdown icon"
        onClick={[Function]}
      />
      <div
        className="menu transition"
      >
        <div
          aria-checked={false}
          aria-selected={false}
          className="item"
          onClick={[Function]}
          role="option"
          style={
            Object {
              "pointerEvents": "all",
            }
          }
        >
          <span
            className="text"
          >
            10
          </span>
        </div>
        <div
          aria-checked={true}
          aria-selected={true}
          className="active selected item"
          onClick={[Function]}
          role="option"
          style={
            Object {
              "pointerEvents": "all",
            }
          }
        >
          <span
            className="text"
          >
            25
          </span>
        </div>
        <div
          aria-checked={false}
          aria-selected={false}
          className="item"
          onClick={[Function]}
          role="option"
          style={
            Object {
              "pointerEvents": "all",
            }
          }
        >
          <span
            className="text"
          >
            50
          </span>
        </div>
        <div
          aria-checked={false}
          aria-selected={false}
          className="item"
          onClick={[Function]}
          role="option"
          style={
            Object {
              "pointerEvents": "all",
            }
          }
        >
          <span
            className="text"
          >
            100
          </span>
        </div>
      </div>
    </div>
  </span>,
  <div>
    <h2
      className="ui header"
    >
      No results.
      <a
        href="#"
        onClick={[Function]}
      >
         Try again?
      </a>
    </h2>
  </div>,
  <div />,
]
`;
